### DISCUSSION TITLE
External Librarian/Control App

### DISCUSSION THREAD

**Discussion by BobanSpasic** [probonopd/MiniDexed#196] [Ideas]
Hi @all,

in a couple of days I'll get my hardware to build the MiniDexed. 
I would also like to write a PC-side (Win/Lin/ QT5) Librarian/Control app.
My question to the developers is - could you include/enable the FTP or TFTP for loading the SysEx files to the SD-Card? 
MT32L has it, and it is also based on Circular.
I would do it myself, but I do not "speak" C/C++.

**Comment by probonopd** [comment]
Welcome  @BobanSpasic. sysex via serial is being worked on: https://github.com/probonopd/MiniDexed/pull/195

You need something like [this](https://github.com/probonopd/MiniDexed/wiki/Hardware#usb-midi-device) to connect MiniDexed to your PC and use an app like Dexed on the PC to send sysex to MiniDexed.

**Comment by BobanSpasic** [comment]
Hi probonopd,
I am talking about a Librarian for sounds, with upload to MiniDexed and renaming the files to the naming convention the MiniDexed needs and such. Also, editing performance.ini etc.
Sending sysex will just edit the current voice(s), or you are thinking about defining some kind of sysex that would also store the voices to the sd-card?

**Reply by probonopd** [reply]
I am assuming that once we can send voice sysex to MiniDexed, someone will ask for being able to store them ;)

**Reply by probonopd** [reply]
They now get stored as part of the Performance.

**Reply by BobanSpasic** [reply]
Not exactly my idea...
With a LAN/WiFi and FTP I could manipulate the files on the SD-Card without a need to get the card out of the Pi and put it in the card-reader on the PC.
I know I am asking too much because you do not have enough developers at the moment. Let's just put this idea on ice

**Reply by now-its-dark** [reply]
Is it possible that a bulk dump operation (aka a complete voice bank transfer) via sysex could trigger an event in minidexed, to store that data as a new .syx file on the SD card? Seems ideal to use preexisting multitimbral dx editors/librarians like[ SynLib DX-TX](http://m.bareille.free.fr/synlibdxtx/synlibdxtx.htm)

**Reply by probonopd** [reply]
> With a LAN/WiFi and FTP

Why not use MIDI sysex, the proven way to do this (and most editor software already supports it)?

> Is it possible that a bulk dump operation (aka a complete voice bank transfer) via sysex could trigger an event in minidexed, to store that data as a new .syx file on the SD card?

Yes, that could be done.

**Reply by BobanSpasic** [reply]
> > With a LAN/WiFi and FTP
> 
> Why not use MIDI sysex, the proven way to do this (and most editor software already supports it)?
> 
1. And how do you edit the performance files? 
On stage, no musician will use pure voices for playing - he uses performances.
See, the poly/mono and portamento are significant parameters, and thus are performances needed more than single voices. A keyboard split would be here also a nice addition (like TG1 plays bass guitar in the range C1 to C3, TG2 plays some solo instrument in the range C#3 to C6). E.g. DX21 have had a Dual and Split modes. MiniDexed has not just Dual, it has an "Octal" mode. Split is still missing.
2. If MIDI IN is occupied by a keyboard, the only possible free connection to PC (Librarian/Editor) is LAN/WiFi

I apologize if my posts are a bit rude - English is not my first language and I didn't learn it in school. I picked it up from the internet and TV.

**Reply by probonopd** [reply]
Thanks for sharing your thoughts, it's always good to brainstorm about how things can be made better.

* Regarding 1: No editor is using/supporting the `performance.ini` format that MiniDexed is using. One would have to write a new one... or just edit the voices via MIDI sysex commands, and then save the resulting performance on the MiniDexed device itself (maybe we could add a sysex command to do that as @[now-its-dark has suggested
* Regarding 2: You can attach as many USB MIDI devices as you like, plus one serial MIDI device

**Reply by now-its-dark** [reply]
@BobanSpasic it was my understanding that these performance parameters you mentioned are defined in the sysex spec for these synths, mostly at least- like poly/mono, portamento, keyboard range per-part and could be editable via sysex as with other voice parameters. Like in the UI for this editor, was this not what you were thinking?
![synlibdxtx1](https://user-images.githubusercontent.com/5728860/170895087-20df7712-fd48-4f72-9c0e-b41de3aecc37.gif)

**Reply by BobanSpasic** [reply]
@now-its-dark 
The Performance parameters are not part of **Voice bulk dump**. There is another SysEx for **Performance bulk dump**, and as far as I understand - it is not supported by MiniDexed.
http://synthmanuals.com/manuals/yamaha/tx7/owners_manual/tx7e1.pdf - read from page 38 onwards

**Reply by BobanSpasic** [reply]
> Thanks for sharing your thoughts, it's always good to brainstorm about how things can be made better.
> 
>     * Regarding 1: No editor is using/supporting the `performance.ini` format that MiniDexed is using. One would have to write a new one... or just edit the voices via MIDI sysex commands, and then save the resulting performance on the MiniDexed device itself (maybe we could add a sysex command to do that as @[now-its-dark has suggested

I was offering to write an editor/librarian in my first post in this discussion.
I am a versed programmer, it is just that I work in Freepascal/Lazarus, not in C++, that's why I can't get involved in development of MiniDexed, but I can still write an Editor/Librarian for MiniDexed.
btw. I wrote the PSSEdit (Editor for 2-OP Yamaha synths) some 25 years ago... in Visual Basic for Windows 95

**Reply by dcoredump** [reply]
Hi @BobanSpasic,

the idea with the (T-)FTP server is not bad in principle, but it is a very unconventional way to adjust the sound data. The problem is that a complete network stack would have to be installed (incl. DHCP). Then there is the (T)FTP server and some form of access control. This must then also co-operate with the other processes (sound production, MIDI data processing, display, Encoder, ...). This is not an easy task.

In principle, the upload of complete bank works. But I don't have any code at this place yet (see [mididevice.cpp#L43](https://github.com/probonopd/MiniDexed/blob/60ab4a562b38cf9ac3e05b85462d61193bddef27/src/mididevice.cpp#L434)). The problem is in fact that you need a mechanism to assign the file bank name. In [MicroDexed](https://codeberg.org/dcoredump/MicroDexed/src/branch/master/UI.hpp#L4800) I have solved it in such a way that one must switch first into the bank transfer mode and assign a name (with the encoders). Then the upload can take place and the bank is stored on the SD card.

If one would build in a complete network stack, one could also think about the use of OSC. But as said: The expenditure is relatively high in my opinion and it is not necessarily the usual musician interface.

Regards, Holger

**Reply by now-its-dark** [reply]
> 

@BobanSpasic Thanks for the clarification. Guess I was incorrect in thinking that the "legacy" / DX-compatible performance params were supported, separately to the extended capabilities in the minidexed performance.ini

**Reply by now-its-dark** [reply]
Hi, I had question following this up- though there are existing editors which can define  many performance parameters for original hardware, that sysex is not used because performance data is defined separately in minidexed, as performance.ini. 
 
So I'm curious how feasible it would be to capture and translate those DX/TX compatible sysex performance params and update the  performance.ini to those values where applicable. 

It wouldn't be a complete solution for editing the wider range of parameters in this file, but would likely reduce the amount of pulling out the SD card quite a bit and maybe more importantly, would allow for a fuller degree of compatibility with the original devices. Sorry if this has already been discussed, I wasn't been able to find a direct reference to it.

**Reply by probonopd** [reply]
Shouldn't the storage format of the file on disk be irrelevant once everything can be configured using MIDI commands?

**Comment by BobanSpasic** [comment]
So, I got some spare time, and I started my work here: https://github.com/BobanSpasic/MiniDexedLibrarian
There isn't much there at the moment, just untested base structures. Loading a sysex file and getting the voice name(s) is working.
Translating the VMEM to VCED formats (packed to expanded) and vice versa is not tested.
The single voice SysEx files from DX7 are in VCED format (one byte pro parameter), and voice bank dumps are in VMEM format (a couple of parameters packed into one byte if the parameters are short enough to be combined).
One useful function is ContainsDXData from untDXUtils, which can identify a lot of different DX dump formats.
All the functions are designed to work with file offset. This way it can extract SysEx data from files generated by various librarian-apps that are adding their own header at the beginning of the file.

It ain't much, but it's honest work 
(google for the phrase if you do not know it already :) )

![MiniDexedLibrarian](https://user-images.githubusercontent.com/68187526/175766407-7330146c-1e52-4de5-ae00-b0d312049cf4.jpg)
In the screenshot is a test app for the units. You can see a bank for DX7II, which contains the basic DX7-compatible data and additional DX7II supplement data.

**Reply by probonopd** [reply]
Interesting! Thanks for sharing @BobanSpasic

**Reply by BobanSpasic** [reply]
There is some progress:
![mdgui](https://user-images.githubusercontent.com/68187526/176009949-f5e8863f-e965-471f-8027-ed4936ab1a1a.png)
Combining voices from different bank files is working (Drag&Drop to the slots)
There is nothing usable for the practical use at the moment, but the base classes are slowly taking a form.

**Reply by probonopd** [reply]
Looking goo!
If I may have one request, please do not use any MiniDexed (or variations of it) logos in the app. Thanks!

**Reply by BobanSpasic** [reply]
Logo removed.

Performance editor:
![mdgui](https://user-images.githubusercontent.com/68187526/176781421-b7183c28-946e-48e6-b726-b72aa2b9bbae.png)

**Comment by BobanSpasic** [comment]
I have released an exe for Windows 64-bit in Bin directory of https://github.com/BobanSpasic/MiniDexedLibrarian
It is the current state of the development.
What is working:
- navigating through DX7 SysEx dumps (bank and single voice)
- drag&drop voices from banks to the empty slots (right pane) and save the resulting bank
- drag&drop voices from banks to the performance slots
- edit performance, load, save, replace voices...

Still not there
- no MIDI at all and thus no MIDI-related operations

MIDI is the next step, and I hope I get it implemented before my vacation in two weeks (no PC/Laptop in vacation - it is a rule... from my wife for me...)

**Comment by BobanSpasic** [comment]
Can MiniDexed receive a voice bulk dump (32 voices)?
Which kind of dumps can MiniDexed receive?

**Reply by arsamus** [reply]
Hi @BobanSpasic, as I can see on the code, right now just voice data dump can be processed but not the whole bank of 32 voices. Minidexed just read and load in RAM banks from SD card but can't save them.

**Reply by BobanSpasic** [reply]
Thanks. I was wondering, because sending bulk dumps to Dexed (standalone Windows version) works.
btw. I renamed the app to MiniDexed Control Center.

**Reply by arsamus** [reply]
Yes, but Desktop Dexed includes Synth Dexed core and complement it with the graphical UI and other control processes. Minidexed do the same and all of those complement processes have to be created because there are not part of synth dexed core.

In case of Bank dumping it is necessary to decide and define where data has to be stored as soon as is received (I think in the immediate empty slot after last loaded bank, RAM is reserved and structured for 127 banks) and then how to store it on SD card folder. Then write the proper code.

**Comment by BobanSpasic** [comment]
Hi folks.
MiniDexed Control Center is still worked on, and it is progressing:
- the interface is redesigned. It isn't fancy anymore, but it is done for the sake of portability
- Linux/GTK port with working MIDI Out
- Database for DX7 Voices (with a structure that will also support DX7II supplement, that can be further converted to MiniDexed performance parameters)
- as for now, banks from DX7II are just partially imported (just the parameters found on DX7)
- file manager for SysEx and Performance files (simplified adding of 00000x part of the file names, drag&drop re-ordering)

What is also important, Control Center is adding some new lines to minidexed.ini:
CommentLine1
CommentLine2
CommentLine3
These lines could be used to write a couple of words about the hardware configuration or anything else.

Control Center is also adding the following lines to the performance.ini files:
Name
Category
Origin

Name is used in Control Center for overriding the file name (else: the filename without the "00000x_"-part is used).
Category - e.g. Keyboards, Synth, Drum, Ethnic, Strings etc. Categories are not pre-defined.
Origin - the name of the author or of the collection, to show some respect to the author.

A new release will follow in a couple of days. I need to get the job done on importing DX7II banks before the release.

**Reply by probonopd** [reply]
Looking forward to it. Any particular reason for choosing Gtk over Qt?

**Reply by BobanSpasic** [reply]
I'll try Qt again. It was really bad-looking with the previous interface design. Perhaps is now better with the re-designed interface.
2nd reason was - to use Qt, one more library (libqt4pas.so - Qt for Pascal) must be installed on the system, where Gtk didn't need anything special.

**Reply by BobanSpasic** [reply]
I've managed to pack some DEB files (GTK and Qt versions) Please test and report if it is working for you.

**Reply by probonopd** [reply]
Thanks @BobanSpasic. It always depends on which desktop you are running an app. On my desktops, I don't use anything Gtk based, so Gtk apps look always totally ugly and out of place. The same is true the other way around if you don't generally use Qt. So good to have both options.

How does one compile the app? I am not using Linux, but FreeBSD.

**Reply by pik33** [reply]
A repository files look like Lazarus was used, so install it - I hope it is possible - https://wiki.freepascal.org/Installing_Lazarus_on_FreeBSD - and open the project (.lpi file)

There is even Lazarus based bare metal environment for RPi called Ultibo.

**Reply by BobanSpasic** [reply]
In Lazarus, first you need to use Online Package Manager from the Package menu, and install the following packages:
    ATShapeLine
    EyeCandyControls
    HashLib4Pascal
Recompile Lazarus when asked.
Now you can open the project file without having Lazarus complaining about the missing packages.
I guess the project will not compile, or it will produce non-functional binary because the IFDEFs for BSD are missing. I have strict IFDEFs for Windows and for Linux. I don't know if it will work with changing IFDEFs to UNIX instead of LINUX.

Additionally, you need SQLite3 and PortMidi libraries installed from the FreeBSD repositories for the binary to work.

I am still trying to install FreeBSD and to get it compiled my self. As for now, I got Gnome installed, but there is no app-menu or anything similar, just a plain desktop.

**Reply by BobanSpasic** [reply]
Binary executable uploaded (compiled on FreeBSD 13 for Qt5).

**Reply by probonopd** [reply]
It works!

![image](https://user-images.githubusercontent.com/2480569/207942721-f913b43e-6b71-422f-b776-1a625e0ff62b.png)

Even MIDI In works after doing:

``` 
sudo pkg install qt5pas alsa-seq-server
sudo service alsa_seq_server enable
sudo service alsa_seq_server start
```

**Comment by BobanSpasic** [comment]
Updated to 0.9.2.27 (Win64, Linux GTK2 and Qt5, FreeBSD Qt5).
Files are in the bin directory:
https://github.com/BobanSpasic/MiniDexedLibrarian/tree/main/bin

I got some time to write some documentation (doc directory):
https://github.com/BobanSpasic/MiniDexedLibrarian/tree/main/doc
In the documentation, file Params (in xlsx and ods formats), you can see my draft for the PCEDx data format. This is what I use internally in the MiniDexedCC.

**Reply by Umby68** [reply]
Hi
The bin directory is not present

**Reply by probonopd** [reply]
Check https://github.com/BobanSpasic/MiniDexedLibrarian/releases

**Comment by BobanSpasic** [comment]
Could someone give a try to the Darwin version?
Is MIDI working?
Does it work out of the box, or it needs some additional preparations?

I only have Catalina in a VM. The compiled binary works for me, but testing it on another Mac is something I would like to know if it is working.
Thanks in advance.

**Reply by probonopd** [reply]
Which macOS version is needed? I could test on Mac OS X 10.6 (Snow Leopard).

**Reply by BobanSpasic** [reply]
I am updating my VM to Monterey. It will take a couple of hours.
The problem is, that brew supports versions from Big Sur upwards, where the Big Sur ports may or may not work. So, Monterey should be a safe ground for portmidi.
My problem with all this is, that the Virtual HDD grew over 65GB. I don't know how to slim it down to some reasonable size. Let see how big it will get after the update to Monterey.

**Comment by BobanSpasic** [comment]
Version 1.0.1.41 is up.
"Auto preview" of the voices is implemented (for voices stored in database). It works with Dexed (PC) and with MiniDexed.
Renaming the voices in database does not change the name on display of the MiniDexed (it will be implemented in the future)
